### DISCUSSION TITLE
NanoPC-T6 Kernel 6.x with modules - build and install?

### DISCUSSION THREAD

**Discussion by belveder79** [MichaIng/DietPi#6619] [Q&A] [answered]
I discovered that there is the 6-series dietpi and used that one - also worked nicely with dd-ing the sd card to the emmc. Doing that we end up with this layout on the EMMC:

```
(parted) p                                                                
Model: MMC A3A551 (sd/mmc)
Disk /dev/mmcblk2: 31.0GB
Sector size (logical/physical): 512B/512B
Partition Table: gpt
Disk Flags: 

Number  Start   End     Size    File system  Name      Flags
 1      8389kB  12.6MB  4194kB               uboot
 2      12.6MB  16.8MB  4194kB               misc
 3      16.8MB  21.0MB  4194kB               dtbo
 4      21.0MB  37.7MB  16.8MB               resource
 5      37.7MB  79.7MB  41.9MB               kernel
 6      79.7MB  113MB   33.6MB               boot
 7      113MB   147MB   33.6MB               recovery
 8      147MB   31.0GB  30.9GB  ext4
```

Because I did not find the modules I need for my stuff in the 5.10 kernel shipped with the image, I'm wondering about the workflow to update the kernel on those devices. From the [Wiki at Friendlyelec](https://wiki.friendlyelec.com/wiki/index.php/NanoPC-T6) it seems that one needs to indeed dd stuff to individual partitions. I also tried updating the kernel by manually apting ```linux-image-arm64``` which indeed installed the kernel, but only in the rootfs. This means that the 5.10 kernel sticks as it seems, raising some questions.

- is it intended to be like that for the time being?
- is a kernel update possible at all with the 6-series dietpi image that can be downloaded?
- I also tried the official Armbian image which was flagged as WIP, but this one had a 6.X kernel in it (not sure if it had the modules however). So is this Version not based on Armbian at all, or did the Dietpi version simply not get a bump to a newer Armbian yet?

I would rather not want to go for Ubuntu or the FriendlyElec distro if there is a way to upgrade/customize the kernel on the Dietpi release. Any help appreciated...

**Comment by MichaIng** [comment, marked as answer]
There is no good mainline Linux support for this SoC yet, so for now we use the FriendlyELEC/Rockchip vendor kernel, which is natively outdated at 5.10.160.

At least we'll switch to Armbian's kernel builds soon, probably with next release, so that the new images will contain a single partition only and the kernel image along with boot configs etc in `/boot`, rather than distributed across those 7 partitions. Once this is done, it is at least easier to build the same with mainline kernel. Armbian does not ship them via APT repo yet, but using their build system they can be built already, at least: https://github.com/armbian/build/blob/main/config/boards/nanopct6.wip

Of course we could also try unmodified upstream Linux directly. A device tree for the NanoPC T6 has been added just for first Linux 6.6 RCs. Sadly R6S is still missing: https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/arch/arm64/boot/dts/rockchip

**Comment by belveder79** [comment]
thx a lot for the quick answer... yes, it is indeed very new and it's a pity that things don't move as fast as we want them to do :-)

I will see if I can get the stuff done somehow with the 5.10 kernel - seems that also the FriendlyCore is at that level... I was just wondering if there is a plan to make the R5 and T6 more like the RockPros, which are currently a bit easier to handle...

**Comment by belveder79** [comment]
Ok, I got my own kernel compiled following the instructions [from FriendlyElec](https://wiki.friendlyelec.com/wiki/index.php/NanoPC-T6#Build_Other_Linux). I successfully flashed it to an SD card (modules to rootfs, kernel and resources to partitions). Booted nicely and does what it is supposed to do.

I did not find documentation about the device tree however. There is a dtbs partition which is not overwritten by default (which I feel is a bit weird). Because there is no device tree overlay in this partition setup, it seems that I have to update the device tree manually and flash it to the partition. Can you point me to some documentation how to update the device tree partition?

**Reply by MichaIng** [reply]
I also never had a closer look into those partitions, what is written where etc. And as we'll switch to Armbian's kernel and bootloader builds, which allow a regular single-partition image with kernel, dtb etc stored as regular files in `/boot`, I'm not keen to invest any more time into this FriendlyELEC setup 😉.
